### PR TITLE
ddl: put the non-reorg jobs back to general queue

### LINKED ISSUE #31460: Adding index blocks the upgrade procedure
## Enhancement

https://github.com/pingcap/tidb/pull/27748 moves all of the `ModifyColumn` jobs to adding index queue. If there is a running "ADD INDEX" job, the upgrade procedures like `upgradeToVer75` can be blocked:

https://github.com/pingcap/tidb/blob/088a2444dff5c2239f47529dd6218cbd0771a943/session/bootstrap.go#L1614-L1623

`ALTER TABLE mysql.user MODIFY COLUMN Host CHAR(255)` is blocked by adding index.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #31460

### What is changed and how it works?

This PR puts the non-reorg jobs back to general queue.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by zimulala — COMMENTED** [review]
Because of the parallel DDL, it is possible that "reorg data" is required at first and not later. This can lead to uncertainty.

**Comment by tangenta** [comment]
@zimulala Since running `ModifyColumn` jobs in these queues does not affect the correctness, I think it is not a big problem if it happens **occasionally**.

Our goal is that non-reorg `ModifyColumn` jobs are executed by add-index queue **in the average cases**.

File: ddl/ddl.go [hunk 1]
```diff
@@ -508,8 +508,7 @@ func (d *ddl) asyncNotifyWorker(job *model.Job) {
 	}
 
 	var worker *worker
-	jobTp := job.Type
-	if admin.MayNeedBackfill(jobTp) {
+	if jobNeedBackFill(job, nil, d.store) {
```
Comment (tangenta) [inline ddl/ddl.go:511]: @zimulala Can we notify both workers regardless of the type of the job?

Comment (zimulala) [reply ddl/ddl.go:511]: If the notification is incorrect, tikv needs to be read once. It is better to inform correctly if you can, but occasionally there is an error is not a big problem

File: ddl/ddl.go [hunk 2]
```diff
@@ -521,6 +520,56 @@ func (d *ddl) asyncNotifyWorker(job *model.Job) {
 	}
 }
 
+func jobNeedBackFill(job *model.Job, t *meta.Meta, s kv.Storage) bool {
```
Comment (zimulala) [inline ddl/ddl.go:523]: s/jobNeedBackFill/jobNeedBackfill ?

File: ddl/ddl.go [hunk 3]
```diff
@@ -521,6 +520,56 @@ func (d *ddl) asyncNotifyWorker(job *model.Job) {
...
+		err := job.DecodeArgs(&newCol, &oldColName)
+		if err != nil {
+			logutil.BgLogger().Error("decode modify column job args failed in jobNeedBackFill",
+				zap.Int64("job ID", job.ID),
+				zap.Error(err))
+			return false
+		}
+		var oldCol *model.ColumnInfo
+		if t == nil {
+			err = kv.RunInNewTxn(context.Background(), s, false,
```
Comment (zimulala) [inline ddl/ddl.go:539]: Could we put it to `asyncNotifyWorker`？ Then we needn't check `t`.

Comment (tangenta) [reply ddl/ddl.go:539]: So there will be at least one more RPC to TiKV. I prefer to notify all the workers.

**Review by zimulala — COMMENTED** [review]
Could we put "job need backfill" info to `DDLReorgMeta`? Then we can verify that the value is fixed, not affected by column info, and does not require a transaction.

File: util/admin/admin.go [hunk 4]
```diff
@@ -123,9 +123,20 @@ func IsJobRollbackable(job *model.Job) bool {
 	return true
 }
 
-// MayNeedBackfill returns whether the action type may need to backfill the data.
-func MayNeedBackfill(tp model.ActionType) bool {
-	return tp == model.ActionAddIndex || tp == model.ActionAddPrimaryKey || tp == model.ActionModifyColumn
+// JobNeedBackfill checks whether a job need to backfill data.
+func JobNeedBackfill(job *model.Job) bool {
+	switch job.Type {
+	case model.ActionAddIndex, model.ActionAddPrimaryKey:
```
Comment (zimulala) [inline util/admin/admin.go:129]: Do we need to handle it like `ActionModifyColumn`?

Comment (tangenta) [reply util/admin/admin.go]: Done

File: parser/model/ddl.go [hunk 5]
```diff
@@ -318,6 +319,14 @@ func (job *Job) FinishDBJob(jobState JobState, schemaState SchemaState, ver int6
 	job.BinlogInfo.AddDBInfo(ver, dbInfo)
 }
 
+func (job *Job) NeedBackfill() bool {
+	if job.ReorgMeta == nil {
```
Comment (zimulala) [inline parser/model/ddl.go:323]: How case like "create table"? It's ReorgMeta is nil.

File: ddl/integration_test.go [hunk 6]
```diff
@@ -79,3 +83,32 @@ func TestDefaultValueInEnum(t *testing.T) {
...
+	})
+	lastJobNeedReorg = false
+	tk.MustExec("alter table t modify column a bigint;")
+	require.False(t, lastJobNeedReorg)
+	lastJobNeedReorg = false
+	tk.MustExec("alter table t modify column b char(255);")
+	require.False(t, lastJobNeedReorg)
+	lastJobNeedReorg = false
+	tk.MustExec("alter table t modify column a varchar(100);")
+	require.True(t, lastJobNeedReorg)
```
Comment (zimulala) [inline ddl/integration_test.go:113]: Please add test for “create table” and "add index"

**Review by zimulala — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Comment by AilinKid** [comment]
```
FAIL: ddl_worker_test.go:1533: testDDLSuite.TestParallelDDL
...
```

**Review by wjhuang2016 — COMMENTED** [review]
Since that we don't need to add `NeedBackFill` to the reorg meta.
For `CancelJobs`, we can know which queue should use by the job's offset, that is, if j >= len(generJobs), then the job should be in the `add index` queue.

**Review by wjhuang2016 — COMMENTED** [review]
We can pass a boolean to CtxVar for `modify column`, and check it in MayNeedBackfill if it's type is `ActionModifyColumn`.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**